compiler/rustc_infer/src/infer/opaque_types.rs

```diff
@@ -396,10 +396,29 @@ impl<'tcx> InferCtxt<'tcx> {
     /// defining scope.
     #[instrument(skip(self), level = "trace", ret)]
     fn opaque_type_origin_unchecked(&self, def_id: LocalDefId) -> OpaqueTyOrigin {
```

Comment:
It makes me uncomfortable that we're calling `opaque_type_origin_unchecked` on an associated type. That's probably wrong...

Reply:
Going to check why is being called.

Reply:
I think this is fixed alternatively by #109198, and should make it so that we don't need to modify `opaque_type_origin_unchecked` at all.

Reply:
Can you tell me how #109198 avoid calling this function?.

Reply:
Given some `-> impl Trait`, let's call that opaque `def-id#1` (`DefKind::Opaque`). We later synthesize a `def-id#2`, which is an associated type (`DefKind::AssocTy`).

In `typeck_with_fallback`, we normalize `ty::Alias(ty::Projection, def-id#2)` in the original trait signature to `ty::Alias(ty::Opaque, def-id#2)` *instead* of `ty::Alias(ty::Opaque, def-id#1)`:

https://github.com/rust-lang/rust/blob/c5c43407603599f4ff2b217cc09be2cc5f39967d/compiler/rustc_ty_utils/src/ty.rs#L271-L272

https://github.com/rust-lang/rust/blob/c5c43407603599f4ff2b217cc09be2cc5f39967d/compiler/rustc_ty_utils/src/ty.rs#L287-L295

That's wrong, because def-id#2 is an associated type.

That passes a signature with `ty::Alias(ty::Opaque, def-id#2)` to `check_fn`, which passes it to `replace_opaque_types_with_inference_vars`. That tries to call `opaque_type_origin` on `def-id#2` (reminder: an associated type) which ICEs.

That is what I fixed in #109198.